src/client/auth/aws.rs

```diff
@@ -0,0 +1,421 @@
+    }
+
+    if !server_second.done {
+        return Err(Error::invalid_authentication_response("MONGODB-AWS"));
+    }
+
+    Ok(())
+}
+
+/// Contains the credentials for MONGODB-AWS authentication.
+#[derive(Debug, Deserialize)]
+struct AwsCredential {
+    #[serde(rename = "AccessKeyId")]
+    access_key: String,
+
+    #[serde(rename = "SecretAccessKey")]
+    secret_key: String,
+
+    #[serde(rename = "Token")]
+    session_token: Option<String>,
+}
+
+impl AwsCredential {
+    /// Derives the credentials for an authentication attempt given the set of credentials the user
+    /// passed in.
+    async fn get(credential: &Credential, http_client: &HttpClient) -> Result<Self> {
+        let access_key = credential
+            .username
+            .clone()
+            .or_else(|| std::env::var("AWS_ACCESS_KEY_ID").ok());
+        let secret_key = credential
+            .password
+            .clone()
+            .or_else(|| std::env::var("AWS_SECRET_ACCESS_KEY").ok());
+        let session_token = credential
+            .mechanism_properties
+            .as_ref()
+            .and_then(|d| d.get_str("AWS_SESSION_TOKEN").ok())
+            .map(|s| s.to_string())
+            .or_else(|| std::env::var("AWS_SESSION_TOKEN").ok());
+
+        let found_access_key = access_key.is_some();
+        let found_secret_key = secret_key.is_some();
+
+        // If we have an access key and secret key, we can continue with the credentials we've
+        // found.
+        if let (Some(access_key), Some(secret_key)) = (access_key, secret_key) {
+            return Ok(Self {
+                access_key,
+                secret_key,
+                session_token,
+            });
+        }
+
+        if found_access_key || found_secret_key {
+            return Err(Error::authentication_error(
+                "MONGODB-AWS",
+                "cannot specify only one of access key and secret key; either both or neither \
+                 must be provided",
+            ));
+        }
+
+        if session_token.is_some() {
+            return Err(Error::authentication_error(
+                "MONGODB-AWS",
+                "cannot specify session token without both access key and secret key",
+            ));
+        }
+
+        if let Ok(relative_uri) = std::env::var("AWS_CONTAINER_CREDENTIALS_RELATIVE_URI") {
+            Self::get_from_ecs(relative_uri, http_client).await
+        } else {
+            Self::get_from_ec2(http_client).await
+        }
+    }
+
+    /// Obtains credentials from the ECS endpoint.
+    async fn get_from_ecs(relative_uri: String, http_client: &HttpClient) -> Result<Self> {
+        // Use the local IP address that AWS uses for ECS agents.
+        let uri = format!("http://{}/{}", AWS_ECS_IP, relative_uri);
+
+        http_client
+            .get_and_deserialize_json(&uri, None)
+            .await
+            .or_else(|_| Err(Error::unknown_authentication_error("SCRAM")))
```

Comment:
here and a few places below use "SCRAM" instead of "MONGODB_AWS" (or whatever string we're using)